src/main/java/org/checkerframework/checker/crypto/CryptoAnnotatedTypeFactory.java

```diff
@@ -74,12 +77,83 @@ public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror
 
         @Override
         public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror a2) {
-            return CryptoAnnotatedTypeFactory.this.BOTTOM;
+            if (AnnotationUtils.areSameByName(a1, BOTTOM)
+                    || AnnotationUtils.areSameByName(a2, BOTTOM)) {
+                return CryptoAnnotatedTypeFactory.this.BOTTOM;
+            }
+            if (AnnotationUtils.areSameByName(a1, UNKNOWNALGORITHMORPROVIDER)) {
+                return a2;
+            }
+            if (AnnotationUtils.areSameByName(a2, UNKNOWNALGORITHMORPROVIDER)) {
+                return a1;
+            }
+            if ((AnnotationUtils.areSameByName(a1, ALLOWEDALGORITHMS)
+                            && AnnotationUtils.areSameByName(a2, ALLOWPROVIDERS))
+                    || (AnnotationUtils.areSameByName(a2, ALLOWEDALGORITHMS)
+                            && AnnotationUtils.areSameByName(a1, ALLOWPROVIDERS))) {
+                return CryptoAnnotatedTypeFactory.this.BOTTOM;
+            }
+
+            // Both a1 and a2 are the same annotation kind.
+            ExecutableElement element =
+                    TreeUtils.getMethod(
+                            AnnotationUtils.annotationMirrorToClass(a1), "value", 0, processingEnv);
+            List<String> a1RegexList =
+                    AnnotationUtils.getElementValueArray(a1, element, String.class);
+            List<String> a2RegexList =
+                    AnnotationUtils.getElementValueArray(a2, element, String.class);
+            ArrayList<String> result = new ArrayList<>();
```

Comment:
Could you use [retainAll](https://docs.oracle.com/javase/8/docs/api/java/util/List.html#retainAll-java.util.Collection-) instead?

Reply:
Sure, thanks!